contracts/V1/StateChannelDiamondProxy/StateSnapshotFacet.sol

```diff
@@ -0,0 +1,230 @@
+pragma solidity ^0.8.8;
+
+import "./StateChannelCommon.sol";
+import "../DataTypes.sol";
+import "./AStateChannelManagerProxy.sol";
+
+contract StateSnapshotFacet is StateChannelCommon {
+
+
+    function updateStateSnapshotWithDispute(
+        bytes32 channelId,
+        ForkMilestoneProof[] memory milestoneProofs,
+        StateSnapshot[] memory milestoneSnapshots,
+        DisputeProof memory disputeProof,
+        ExitChannelBlock[] memory exitChannelBlocks
+    ) external onlySelf {
+        // resolve genesis state snapshot source
+        // - if stateSnapshot(on-chain).forkCnt == disputeProof.forkCnt, then the genesis state snapshot is the on-chain stateSnapshot
+        // - otherwise, the dispute is validated and the genesis state snapshot is disputeProof.outputStateSnapshot
+        StateSnapshot memory genesisStateSnapshot = _resolveGenesisSnapshot(
+            channelId,
+            disputeProof
+        );
+
+        _updateStateSnapshot(
+            channelId,
+            milestoneProofs,
+            milestoneSnapshots,
+            exitChannelBlocks,
+            genesisStateSnapshot
+        );
+    }
+
+    function updateStateSnapshotWithoutDispute(
+        bytes32 channelId,
+        ForkMilestoneProof[] memory milestoneProofs,
+        StateSnapshot[] memory milestoneSnapshots,
+        ExitChannelBlock[] memory exitChannelBlocks
+    ) external onlySelf {
+        require(
+            getDisputeLength(channelId) == getSnapshotForkCnt(channelId),
+            "Dispute proof is required"
+        );
+        StateSnapshot memory genesisStateSnapshot = stateSnapshots[channelId];
+
+        _updateStateSnapshot(
+            channelId,
+            milestoneProofs,
+            milestoneSnapshots,
+            exitChannelBlocks,
+            genesisStateSnapshot
+        );
+    }
+
+    function _resolveGenesisSnapshot(
+        bytes32 channelId,
+        DisputeProof memory disputeProof
+    ) internal view returns (StateSnapshot memory) {
+        if (getDisputeLength(channelId) != getSnapshotForkCnt(channelId)) {
+            // dispute proof is required
+            Dispute memory dispute = disputeProof.dispute;
+            uint disputeTimestamp = disputeProof.timestamp;
+
+            require(
+                _isDisputeProofProvided(disputeTimestamp),
```

Comment:
why is this check needed? If dispute proof is not provided the _isDisputeCommitmentValid bellow would anyway return false